### PR TITLE
Ordered quantification

### LINKED ISSUE #366: Instantiation arity mismatch
The following program causes an `Instantiate.ArityMismatch` error
```
fun f(p, q) {
  switch (0) {
    case 0 -> p(q)
    case n -> f(p, fun(_) { true })
  }
}
```
The error message is
```
# Type variables (total 0)


# Type arguments (total 2)
1. `Type (`MetaTypeVar (...))
2. `Row (({}, ..., false))
Arity mismatch in type application
Expression: `TAppl (((`Var ("q"), ...),
         [`Type (`MetaTypeVar (...)); `Row (({}, ..., false))]))
Type: (_) -> Bool

```

The error does not happen when using `if-then-else` rather than the `switch`, e.g.
```
fun f(p, q) {
  if (true) {
    p(q)
  } else {
    f(p, fun(_) { true })
  }
}
```

### LINKED ISSUE #597: Desugaring session exceptions is not type-preserving
Desugaring session exceptions - specifically the part that inserts top-level handlers inside a `spawn`  breaks typing. The problem is that the translation to insert top-level handlers creates a `TryInOtherwise` whose body needn't have the special `_SessionFail` effect. (This breaks the typing of practically every single program that uses session exceptions as session typing with spawning is rather useless.)

A couple of possible fixes come to mind.

 * One option would be to change the types of the family of spawn functions such that the body always has the special `_SessionFail` effect whenever the `--session-exceptions` flag is set.

 * Another (probably better) option would be to insert an upcast on the body during the translation to enable top-level handlers.

### PR DESCRIPTION
**Pull request by SquidDev**
This PR is a bit of a behemoth, for which I apologise. This branch has rather become a melting pot of various changesets...

### Improvements to quantifier type checking
I'm least familiar with these changes, so please shout if there's something obvious that I've missed!

 - Remove the idea of flexible quantifiers. They can still technically be expressed right now, but should never actually occur.
 - Require that the order of quantified variables within a forall is consistent.
 - Remove quantified instantiation (fixes #366).
 - Prohibit free variables on quantified signatures (`forall a. (a) -e-> a` is no longer valid).

### Various desugaring improvements
There's been several changes to try to make desugaring as correct and type-preserving as possible. While it is still not perfect (I've some additional changes on a separate branch, and there's still some issues with generalisation and CP), this does substantial progress in improving the situation.

There's been quite a lot of changes with regards to this, so but the most important are probably as follows:

 - Improvements to desugaring of session handlers, fixes #597 (previously #621).
 - Allow writing "unsafe signatures" for various prelude methods (#649). This allows us to force wild functions (namely `map`, `concatMap`, etc...) to have a non-wild signature. 
   This also removes the patching of the prelude we were previously doing.
 - Re-use the previously inferred type of functions when running the type checker on subsequent passes. This is required for as `fun f()() { f()() }` will have a different type to `fun f() { fun() { f()() } }`, and `DesugarFuns` would otherwise cause programs to no longer type check.
 - Effect row variables now have an `(Unl, Any)` subkind by default, rather than `(Any, Any)`.
 - When `recheck_frontend_transformations` and `debug` is enabled, we'll also print the name of the pass which introduced an error.
 - Add `recheck_frontend_transformations_dump` flag. When this and `debug` is enabled, we will dump the AST from before and after this transformation pass, allowing you to compare them.
 - Add `recheck_frontend_transformations_filter` option. This can be a comma-separated list of passes you want to recheck. Obviously if a previous pass fails, you'll still get an error, just later on.

### FreezeML
This implements most of FreezeML (the monomorphism restriction on arguments is not implemented).

 - Frozen variables are written using the tilde (`~id` for instance), and will not be instantiated during type checking. Otherwise they are treated the same as normal variables.
 - An explicit generalisation operator `$`. For instance, `$(fun(x) { x })` produces a term of type `forall a::Any,b::Row. (a) -b-> a`.
 - An explicit instantiation operator `@`.

### CODE REVIEW

**Comment by atennapel** [comment]
Could you give a reference on "FreezeML", I would like to read about it but not getting any results on Google.

**Comment by slindley** [comment]
There's nothing on the web because we haven't yet finished writing the paper. There will be a talk at TyDe 2019 along with a two page extended abstract.

File: tests/freezeml.tests [hunk 1]
```diff
@@ -0,0 +1,204 @@
+Polymorphic Intantiation (1)
```
Comment (SimonJF) [inline tests/freezeml.tests]: nit: typo here

File: tests/polymorphism.tests [hunk 2]
```diff
@@ -93,25 +93,33 @@ Quantifiers in different orders
...
 
+Polymorphic function arguments and freeze
+fun poly(f : (forall a::Any,e::Row.(a) -e-> a)) {(f(42), f(true))} poly(~id)
+stdout : (42, true) : (Int, Bool)
+
 Unwrap polymorphism
 sig ids : [forall a,b::Row.(a) -b-> a] var ids = [fun (x) {x}]; hd(ids)(3)
 stdout : 3 : Int
 args : --config=tests/enable_quantified_instantiation.config
+ignore : quantified instantiation no longer available
```
Comment (SimonJF) [inline tests/polymorphism.tests]: If it's no longer available, should we make it a failing test case?

**Comment by SimonJF** [comment]
Also, please add

"
- Remove prelude patching
- Add flags for checking frontend transformations after each step
"
to the patch notes, preferably detailing the new flags. Thanks!

File: core/sugarTraversals.ml [hunk 3]
```diff
@@ -693,6 +686,45 @@ class map =
...
+      let a = o#binder a in
+      let d = o#funlit d in
+      let e = o#location e in
+      let f = o#option (fun o -> o#datatype') f in
+      { fun_binder = a;
+        fun_linearity = b;
+        fun_definition = (c, d);
+        fun_location = e;
+        fun_signature = f;
+        fun_unsafe_signature = g; }
```
[the remark this answers is not in the document]
Comment (SquidDev) [reply core/sugarTraversals.ml]: Ahh - the unsafe signature is just a boolean now, specifying whether the signature is unsafe or not. Thus we don't need to visit it any more.

File: core/parser.mly [hunk 4]
```diff
@@ -488,6 +497,7 @@ qualified_type_name:
 atomic_expression:
 | qualified_name                                               { with_pos $loc (QualifiedVar $1) }
 | VARIABLE                                                     { with_pos $loc (Var          $1) }
+| TILDE VARIABLE                                               { with_pos $loc (FreezeVar    $2) }
```
[the remark this answers is not in the document]
Comment (SquidDev) [reply core/parser.mly]: Yes. It doesn't cause problems in the parser - freezing occurs in expressions, and session dualisation is part of datatypes, so there's no chance of conflicts. However, I can definitely see that there's potential cause for confusion.

I'm not really sure of a nice solution - there's a limited number of glyphs, and I think Sam settled on `~` as the best choice.

File: core/desugarInners.ml [hunk 5]
```diff
@@ -11,18 +11,27 @@ open Sugartypes
...
-        ~message:"Mismatch in number of quantifiers and type arguments") in
-  go extras
+(* FIXME: replace failwith *)
+
+let rec add_extras qs (extras, tyargs) =
+  match qs, extras with
+  | [], [] -> []
+  | q::qs, None::extras -> Types.type_arg_of_quantifier q :: add_extras qs (extras, tyargs)
+  | _::qs, Some i::extras -> List.nth tyargs i :: add_extras qs (extras, tyargs)
+  | _, _ -> failwith "Mismatch in number of quantifiers and type arguments"
```
[the remark this answers is not in the document]
Comment (SquidDev) [reply core/desugarInners.ml]: I've added this as an `InternalError` for now - if we ever hit this, the type checker is rather broken. Happy to add a new exception to `Errors` if preferred though.

File: core/parser.mly (hunk 4 above)
[the remark this answers is not in the document]
Comment (SimonJF) [reply core/parser.mly]: Ah, I'd missed that it was an expression. Sure, that's fine, thanks for the clarification.

File: core/sugarTraversals.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (SimonJF) [reply core/sugarTraversals.ml]: Great, thanks for the clarification.

File: core/desugarInners.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (SimonJF) [reply core/desugarInners.ml]: Fine by me.

**Comment by SimonJF** [comment]
Thanks. Let's merge. Congratulations, and thanks for all the hard work.

**Pull request merged**